### ISSUE TITLE
Create ESLint plugin for CSS in JS?

### ISSUE THREAD

**Issue by ismay** [paypal/glamorous#28]
Great library! From the readme: 

> There are actually quite a few solutions to the general problem of styling in React. This isn't the place for a full-on comparison of features, but I'm unaware of any which supports all of the features which this library supports.

One thing that styled-components does have is a [stylelint processor](https://github.com/styled-components/stylelint-processor-styled-components), which allows you to use [stylelint](https://github.com/stylelint/stylelint) (_the_ css linter) for styles written with styled-components. Which is pretty nice to have.

Just a suggested feature. Think it'd make this library even more complete! (See also [this article](http://mxstbr.blog/2016/12/linting-styles-in-js-with-stylelint/))

**Comment by kentcdodds** [comment]
Agreed! Let's see if [this](https://twitter.com/kentcdodds/status/850371960928219136) turns up anything. I know that a LOT of CSS-in-JS solutions would benefit from a linter like this.

If it doesn't exist, would you care to create it? If you've not made an ESLint plugin before or worked with ASTs before, you can learn about it from [my talk](https://youtu.be/CFQBHy8RCpg?list=PLV5CVI1eNcJgNqzNwcs4UKrlJdhfDjshf) :)

**Comment by ismay** [comment]
> If it doesn't exist, would you care to create it? If you've not made an ESLint plugin before or worked with ASTs before, you can learn about it from my talk :)

I wouldn't mind. I'd then make a stylelint processor for it, based on Max Stoiber's remarks on creating the same for styled-components:

> The first idea was to leverage the power of eslint, the JavaScript linter, to parse the JavaScript and then run stylelint from there on the extracted CSS string. A lot of people already use and rely on eslint, so providing an eslint rule that does this would make it easy to adopt and use.

> Sadly, that doesn’t work. After building a first prototype I noticed that for some unknown reason the stylelint results weren’t coming back to the CLI. I was executing stylelint on the right code, but the user couldn’t see the result! I asked around and somebody in the eslint Gitter chatroom explained to me why:

> eslint rules have to be sync, but stylelint has to be run async.

It might be quite difficult, but I'll see what I can do if I have some time!

**Comment by kentcdodds** [comment]
Cool! Just ping us here once you start working on it. Until then, anyone else can feel free to try to do this :)

**Comment by ismay** [comment]
> Cool! Just ping us here once you start working on it.

I will! 👍

**Comment by ismay** [comment]
One thing I wanted to ask (since I've not made anything like this before): where do the styles defined with glamorous get parsed to css? Because the way I see it the processor would basically do this:

1. Processor checks if there is a `glamorous` import
2. Checks for glamorous components
3. Gets their defined styles
4. Parses them to css (with sourcemaps)
5. Lint resulting css (and ignore any warnings that can't be influenced by the user, like missing semicolons, etc.)
5. Returns any valid warnings (mapped to original lines with sourcemaps)

So what I'm wondering is how to do step 4 in the same way that it'll be done for the actual component.

**Comment by kentcdodds** [comment]
Good question. Actually I wouldn't try to convert it to CSS at all. I would just write the tool from scratch and look for the properties in the object and have special rules for each property. Can't do too much about dynamic styles, but for every property that's a literal style you could check statically.

**Comment by ismay** [comment]
> Actually I wouldn't try to convert it to CSS at all.

If we want to use stylelint we'll have to though. But I think it's worth it, as it's quite a rich ecosystem for css linting so once we're tapped into that we don't have to do much else (see also this [tweet](https://twitter.com/stylelint/status/788795790433001472?ref_src=twsrc%5Etfw&ref_url=http%3A%2F%2Fmxstbr.blog%2F2016%2F12%2Flinting-styles-in-js-with-stylelint%2F)).

Also, I think it'd yield more accurate results and hopefully be easier to maintain, as you're using the same css that the browser eventually uses and you don't have to maintain a custom parser. And yeah, nothing we can do about dynamic styles, those'll just have to be dropped probably.

But I'll see where I get with just the processor first, will probably be easier to reason about with a working example. I'll post here if I have something.

**Comment by kentcdodds** [comment]
I can see the value in doing it that way and reusing the work that's been done in stylint. I wonder whether stylint exposes validators for each individual property 🤔 . The thing is, generating the CSS will be a challenge, but you can look into how https://github.com/threepointone/glamor accomplishes it...

**Comment by ismay** [comment]
I've created a simple sketch to get some discussion started on how we could tackle this:

* [stylelint-processor-glamorous](https://github.com/ismay/stylelint-processor-glamorous)
* [travis build](https://travis-ci.org/ismay/stylelint-processor-glamorous/builds/221827041#L211)

The failing test is the point where we diverge from how [stylelint-processor-styled-components](https://github.com/styled-components/stylelint-processor-styled-components) does the processing. Because styled-components uses css (like) syntax they have a ready-to-lint css string, where we still need to convert from a js object to a css string.

> you can look into how https://github.com/threepointone/glamor accomplishes it...

I've looked into how glamor parses styles, and it seems to be documented here:

* https://github.com/threepointone/glamor/blob/master/docs/implementation.md
* https://github.com/threepointone/glamor/blob/master/docs/implementation.md#generate-css

Basically:

> we use a vendored version of React's CSSPropertyOperations to convert the object into a css string

It would be nice if `CSSPropertyOperations ` could somehow be used to convert our styles to a css string. Besides parsing the js to css we'll have to be able to catch all of the styles in which a glamorous component can be defined. Since I haven't dealt with parsing an AST before I'm not sure how much of a challenge that'll be.

Looking forward to some feedback and ideas on this. I think that this would be very reusable for other css-in-js libs, which would basically unlock the stylelint ecosystem for all of them as well. So anyone who wants to help or give feedback is very welcome :)!

**Comment by kentcdodds** [comment]
Thanks for the work on this so far and thanks for the update. I can appreciate the value of leveraging all the work that's gone into stylelint. At the same time, if I were to do it myself, I would write a normal ESLint plugin that validates things via the normal AST. So I'm unsure of whether I can personally help you out with the stylelint-processor-glamorous project.

But maybe someone else can help you out.

**Comment by kentcdodds** [comment]
Here's the basic idea I'm thinking of: [astexplorer.net](https://astexplorer.net/#/gist/00cffc2e3c5280b68e4ef58debedb228/8e8abc7c0523dc4e586328ba99fadbf64024a5c8).

Pretty much all that we need to do on top of that is:

- [ ] add validators for every property (not a small effort, but you know)...
- [ ] scope the checking to only those objects being passed to the `glamorous` function
- [ ] write tests
- [ ] publish the module

I think this would be much more straightforward (and generally applicable) than trying to generate the CSS and passing it to stylelint.

**Comment by ismay** [comment]
> Thanks for the work on this so far and thanks for the update. I can appreciate the value of leveraging all the work that's gone into stylelint. At the same time, if I were to do it myself, I would write a normal ESLint plugin that validates things via the normal AST

I can understand that. I have a bit of a vested interest in getting it to work with stylelint since it has some plugins that I don't really want to drop. Please let me know if you want to keep the discussion in this issue limited to just an eslint plugin.

However, there's some good news, I found out that stylelint supports [custom syntaxes](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/node-api.md#customsyntax), and that there is already a plugin that can parse css-in-js syntax: [postcss-js](https://github.com/postcss/postcss-js) listed [here](https://github.com/postcss/postcss#syntaxes). So we don't have to handle the parsing ourselves it seems 🎉

Do you think that would make it a more feasible solution? At this point it should be pretty similar to implementing an eslint plugin, with the benefit of not having to write any rules ourselves. As I said, I have kind of a vested interest, so I'll leave the decision to you.

**Comment by kentcdodds** [comment]
Hey @ismay, you can definitely feel free to proceed with what you'd like to. If it really works well, then I'm happy to promote it! I don't have time to work on my idea right now anyway :)

**Comment by ismay** [comment]
I've worked on this some more, but hit a dead end. Stylelint doesn't currently support async processors (which I need to parse the css-in-js), so I think my attempt isn't going to work for now. So I think that your solution is definitely the only feasible option.

**Comment by kentcdodds** [comment]
I'm sorry to hear that :-( Anyone can feel free to build on what I started :) It would really be a great thing for the community and would probably get quite popular :)

**Comment by jackyho112** [comment]
@kentcdodds 

May I give this one a go? I will have to watch that video about your talk on AST and look into how to write a Babel plugin/linter. If that's cool with you, I will get started. 😄

**Comment by kentcdodds** [comment]
For sure!

**Comment by kwelch** [comment]
Should it lint every object expression? Wouldn't it better to find based on object being called? I guess step one is to make it work against all object expressions and work toward the utilized ones.

**Comment by kentcdodds** [comment]
Correct @kwelch, as one of the noted TODOs above:

> scope the checking to only those objects being passed to the glamorous function

**Comment by paulmolluzzo** [comment]
> add validators for every property (not a small effort, but you know)...

This package could help with creating validators: https://github.com/a-axton/css-properties-values

**Comment by kentcdodds** [comment]
OH SNAP @paulmolluzzo! THAT IS SO AWESOME! THANK YOU SO MUCH FOR SHARING!!!

I hadn't expected such a package would exist. Interestingly here's the value for `background`:

```
{
  "property": "background",
  "values": [
    "background-color",
    "background-image",
    "background-position",
    "background-size",
    "background-repeat",
    "background-origin",
    "background-clip",
    "background-attachment",
    "initial",
    "inherit"
  ]
}
```

I'm not sure whether those are actually acceptable values, but I _think_ that while we're checking the `values` array, we'll need to see whether the `value` actually has a `property` in the package and if it does then we can use those values to validate the whole `background` property. It almost seems like this is a primative up which to build a `is-valid-css-value` package (which unfortunately does not exist). Anyone wanna build that!?

**Comment by paulmolluzzo** [comment]
> It almost seems like this is a primative up which to build a is-valid-css-value package (which unfortunately does not exist). Anyone wanna build that!?

Lemme see if I can whip something up.

I was poking at a full `babel`/`eslint` plugin but don't fully grok how to build the plugin and test it.

**Comment by philpl** [comment]
I think we can build a really neat and efficient data structure from `css-properties-values`. Took a quick look at it and a lot of it is not normalised or in a usable format.

I'll try to write a babel-based script to convert it to a smaller js file with integer based constants. This way we might be able to ship it client-side for things like live-editor linting. :smile:

**Comment by philpl** [comment]
There's a major problem with the data from `css-properties-values` unfortunately. They don't differentiate between allows values for normal properties and enumerated values for shorthands 😢 

So for example:

```js
  {
    property: "box-shadow",
    values: [
      "none",
      "h-shadow",
      "v-shadow",
      "blur",
      "spread",
      "color",
      "inset",
      "initial",
      "inherit"
    ]
  }
```

We'd need this to be an array of arrays to work, e.g:

```js
  {
    property: "box-shadow",
    values: [
      "none",
      [
        "h-shadow",
        "v-shadow",
        "blur",
        "spread",
        "color",
        "inset"
      ],
      "initial",
      "inherit"
    ]
  }
```

And some of these in the inner array are even optional 💫 

Does anyone know of a more accurate data set?

**Comment by paulmolluzzo** [comment]
Got a module started with `css-properties-values` as the base: https://github.com/paulmolluzzo/is-valid-css-value

> They don't differentiate between allows values for normal properties and enumerated values for shorthands 😢

@philpl is right, lots of properties don't have all the enums we'd want to have. `color` is a glaring example:

```js
{"property":"color","values":["color","initial","inherit"]}
```

**Comment by kentcdodds** [comment]
> This one is MDN's data set,

lol, that's huge. Glad we don't have to include _that_ in glamorous or styled-components! Yeah, I think we should use that one. If we could create an `is-valid-css-attribute` package based on that, I think that'd be a super awesome win!

**Comment by lydell** [comment]
https://github.com/ben-eb/css-values/ might be worth taking inspiration from.

**Comment by paulmolluzzo** [comment]
> https://github.com/ben-eb/css-values/ might be worth taking inspiration from.

From that README:

> We support 254 of 361 CSS properties (70.36%).

**Comment by jackyho112** [comment]
Finally got some more time to dig deeper. I was trying to validate CSS values by parsing property syntaxes on https://github.com/mdn/data/blob/master/css/properties.json. But it turns out to be more difficult than I thought. I guess I should not be surprised that doing what https://github.com/ben-eb/css-values/  is trying to accomplish in a couple of files is not an easy task 😄 .

I think I will continue with that in a separate branch. So far the plugin will only validate the property names. And it will only validate the object literals with at least one valid CSS property in it. 

I experimented with identifying the relevant `glamorous` object but found that to be more complicated than I thought especially given the amount of ways an object can be passed in. Plus, to do that, I would also like to account for an object that's imported from a separate file(I think ESLint can do that though).  

Anyway, I will put it together without validating values for now and then open it for people's help.

**Comment by itsandrewsmith** [comment]
hi, I'm not sure if this is something that is still up for grabs -- it looks like someone has already worked on it a bit

either way, I wanted to try it out after watching some (lots) of your talks Kent. I've basically just bootstrapped the css-values package, so its still missing validation for properties. but its a start -- and my first package! woo!

https://www.npmjs.com/package/eslint-plugin-css-js

**Comment by kentcdodds** [comment]
This is awesome! Thanks for working on this @itsandrewsmith! Let's move discussion over to that repo :clap: I'm excited to try it out.

Maybe someone could make a PR to add it to this project to lint our tests 🤔

**Comment by itsandrewsmith** [comment]
thanks Kent! any feedback / suggestions are welcome.

**Comment by alp82** [comment]
Here is another great library that might help with linting/validation: https://github.com/mahirshah/css-property-parser

**Comment by zabute** [comment]
I just released [stylelint-processor-glamorous](https://github.com/zabute/stylelint-processor-glamorous). I'm actually still new to glamorous so I don't have any large codebases to test it against. It would be great if I could get some feedback.

I'm loving glamorous, btw, and I want to thank the contributors for all the work thats gone into it. :+1:

**Comment by kentcdodds** [comment]
Well that's awesome! Why don't you go ahead and add a reference to it in the README!

And feel free to [add yourself](../blob/master/CONTRIBUTING.md#add-yourself-as-a-contributor) to the [contributors table](../#contributors) with a tool contribution. Thanks so much. This is very cool :+1:

**Comment by paulmolluzzo** [comment]
> I'm actually still new to glamorous so I don't have any large codebases to test it against. It would be great if I could get some feedback.

This is awesome @zabute! 🎉 

If you want to test this against a large codebase, maybe try the [glamorous-website repo](https://github.com/kentcdodds/glamorous-website)?

**Comment by zabute** [comment]
@kentcdodds Just made #385 

@paulmolluzzo Thanks for that tip. The tool is much better now after testing against the galmorous website as it surfaced quite a few bugs I suspected were lurking in there.

Thank you both for the quick responses!!
